text/0000-string-copy_cpointer-copy_cstring-returns-iso.md

```diff
@@ -0,0 +1,111 @@
+- Feature Name: Modify String.copy\_cpointer() and String.copy\_cstring() to return iso^
+- Start Date: 2025-09-10
+- RFC PR:
+- Pony Issue:
+
+# Summary
+
+The two current constructors `String.copy_cpointer()` and `String.copy_cstring()` both take a `Pointer[U8] box` and return a `String ref^`. Given that these two constructors make copies, returning a `String iso^` would be preferable and safe.
+
+Note: In the interests of clarity of writing, I will use `String.copy_cstring()` throughout the rest of this RFC. All of the commentary below should be read to apply to both `String.copy_cstring()`, and `String.copy_cpointer()`.
+
+# Motivation
+
+In many cases, as a library author I want to provide my end-users a mutable and sendable copy of a String received via C-FFI. In other words, a `String iso^`. With the way that the String library is currently written, there are two ways to do this:
+
+## Create a String ref^ using String.from\_cstring() or from\_cpointer() and clone().
+
+```pony
+  var str: String iso = String.from_cstring(ptr).clone()
+```
+
+The main issue with this is that it directly violates the documentation which states: "This must be done only with C-FFI functions that return pony\_alloc'd character arrays".
```

Comment:
side note: that documentation should probably be updated as if it doesn't state a reason. to be safe, it must be assured that the allocator of the memory won't free it for as long as we hold on to it.

what should really be said is that it should only be done with a pointer to memory that we can take ownership of.